The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12242781197/job/34150868939) [(plain)](https://github.com/rust-lang/rust/commit/8f4622d72d034e1a2e36b6e22dd487e01fa448bd/checks/34150868939/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
error[E0261]: use of undeclared lifetime name `'a`
   --> compiler/rustc_codegen_ssa/src/codegen_attrs.rs:795:29
    |
791 | struct MixedExportNameAndNoMangleState {
    |                                       - help: consider introducing lifetime `'a` here: `<'a>`
...
795 |     no_mangle_attr: Option<&'a ast::Attribute>,
    |                             ^^ undeclared lifetime
error[E0425]: cannot find value `no_mangle_attr` in this scope
   --> compiler/rustc_codegen_ssa/src/codegen_attrs.rs:824:83
    |
    |
824 |                     no_mangle_attr: rustc_ast_pretty::pprust::attribute_to_string(no_mangle_attr),
    |
help: you might have meant to use the available field
    |
    |
824 |                     no_mangle_attr: rustc_ast_pretty::pprust::attribute_to_string(self.no_mangle_attr),

   Compiling rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error[E0308]: mismatched types
   --> compiler/rustc_codegen_ssa/src/codegen_attrs.rs:123:25
---
    |
note: method defined here
   --> compiler/rustc_codegen_ssa/src/codegen_attrs.rs:803:8
    |
803 |     fn track_no_mangle(&mut self, span: Span, hir_id: HirId, attr_name: String) {


error[E0609]: no field `no_mangle_attr_name` on type `&mut MixedExportNameAndNoMangleState`
    |
    |
806 |         self.no_mangle_attr_name = Some(attr_name);
    |
help: a field with a similar name exists
    |
806 |         self.no_mangle_attr = Some(attr_name);
806 |         self.no_mangle_attr = Some(attr_name);
    |              ~~~~~~~~~~~~~~

error[E0026]: struct `MixedExportNameAndNoMangleState` does not have a field named `no_mangle_attr_name`
    |
815 |             no_mangle_attr_name: Some(no_mangle_attr_name),
    |             ^^^^^^^^^^^^^^^^^^^
    |             |
    |             |
    |             struct `MixedExportNameAndNoMangleState` does not have this field
    |             help: a field with a similar name exists: `no_mangle_attr`
Some errors have detailed explanations: E0026, E0261, E0308, E0425, E0609.
For more information about an error, try `rustc --explain E0026`.
error: could not compile `rustc_codegen_ssa` (lib) due to 5 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

--- a/compiler/rustc_codegen_ssa/src/codegen_attrs.rs
+++ b/compiler/rustc_codegen_ssa/src/codegen_attrs.rs
@@ -120,7 +120,7 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
                     mixed_export_name_no_mangle_lint_state.track_no_mangle(
                         attr.span,
                         tcx.local_def_id_to_hir_id(did),
-                        rustc_ast_pretty::pprust::attribute_to_string(attr),
+                        attr,
                     );
                 } else {
                     tcx.dcx()
@@ -788,22 +788,22 @@ fn check_link_name_xor_ordinal(
 }
 
 #[derive(Default)]
-struct MixedExportNameAndNoMangleState {
+struct MixedExportNameAndNoMangleState<'a> {
     export_name: Option<Span>,
     hir_id: Option<HirId>,
     no_mangle: Option<Span>,
-    no_mangle_attr_name: Option<String>,
+    no_mangle_attr: Option<&'a ast::Attribute>,
 }
 
-impl MixedExportNameAndNoMangleState {
+impl<'a> MixedExportNameAndNoMangleState<'a> {
     fn track_export_name(&mut self, span: Span) {
         self.export_name = Some(span);
     }
 
-    fn track_no_mangle(&mut self, span: Span, hir_id: HirId, attr_name: String) {
+    fn track_no_mangle(&mut self, span: Span, hir_id: HirId, attr_name: &'a ast::Attribute) {
         self.no_mangle = Some(span);
         self.hir_id = Some(hir_id);
-        self.no_mangle_attr_name = Some(attr_name);
+        self.no_mangle_attr = Some(attr_name);
     }
 
     /// Emit diagnostics if the lint condition is met.
@@ -812,7 +812,7 @@ impl MixedExportNameAndNoMangleState {
             export_name: Some(export_name),
             no_mangle: Some(no_mangle),
             hir_id: Some(hir_id),
-            no_mangle_attr_name: Some(no_mangle_attr_name),
+            no_mangle_attr: Some(no_mangle_attr),
         } = self
         {
             tcx.emit_node_span_lint(
@@ -821,7 +821,7 @@ impl MixedExportNameAndNoMangleState {
                 no_mangle,
                 errors::MixedExportNameAndNoMangle {
                     no_mangle,
-                    no_mangle_attr: no_mangle_attr_name,
+                    no_mangle_attr: rustc_ast_pretty::pprust::attribute_to_string(no_mangle_attr),
                     export_name,
                     removal_span: no_mangle,
                 },

